processing/src/main/java/org/apache/druid/segment/join/HashJoinSegment.java

```diff
@@ -104,16 +105,37 @@ public QueryableIndex asQueryableIndex()
   }
 
   @Override
-  public StorageAdapter asStorageAdapter()
+  public CursorFactory asCursorFactory()
   {
-    return new HashJoinSegmentStorageAdapter(
-        baseSegment.asStorageAdapter(),
+    return new HashJoinSegmentCursorFactory(
+        baseSegment.asCursorFactory(),
         baseFilter,
         clauses,
         joinFilterPreAnalysis
     );
   }
 
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T> T as(Class<T> clazz)
+  {
+    if (CloseableShapeshifter.class.equals(clazz)) {
+      return (T) new CursorFactoryRowsAndColumns(asCursorFactory());
+    } else if (TimeBoundaryInspector.class.equals(clazz)) {
+      return (T) WrappedTimeBoundaryInspector.create(baseSegment.as(TimeBoundaryInspector.class));
+    } else if (TopNOptimizationInspector.class.equals(clazz)) {
+      // if the baseFilter is not null, then rows from underlying cursor can be potentially filtered.
+      // otherwise, a filtering inner join can also filter rows.
+      return (T) new SimpleTopNOptimizationInspector(
+          baseFilter != null || clauses.stream().anyMatch(
+              clause -> clause.getJoinType() == JoinType.INNER && !clause.getCondition().isAlwaysTrue()
```

Comment:
I think this is a pre-existing issue, but, this check should be:

```
!clause.getJoinType().isLefty() && !clause.getCondition().isAlwaysTrue()
```

Because `JoinType.RIGHT` can also filter rows from the base.

Reply:
fixed